src/compiler/crystal/macros.cr

```diff
@@ -1453,11 +1453,39 @@ module Crystal::Macros
     end
   end
 
-  # class Rescue < ASTNode
-  # end
+  # A `rescue` clause inside an exception handler.
+  class Rescue < ASTNode
+    # Returns this `rescue` clause's body.
+    def body : ASTNode
+    end
 
-  # class ExceptionHandler < ASTNode
-  # end
+    # Returns this `rescue` clause's exception types, if any.
+    def types : ArrayLiteral | NilLiteral
+    end
+
+    # Returns the variable name of the rescued exception, if any.
+    def name : MacroId | Nop
+    end
+  end
+
+  # A `begin ... end` expression with `rescue`, `else` and `ensure` clauses.
+  class ExceptionHandler < ASTNode
+    # Returns this exception handler's main body.
+    def body : ASTNode
+    end
+
+    # Returns this exception handler's `rescue` clauses, if any.
+    def rescues : ArrayLiteral(Rescue) | NilLiteral
```

Comment:
1. Why doesn't just returning an empty array suffice?
2. Why `NilLiteral` and not `Nop` like elsewhere?

Reply:
This implementation was based on `Generic#type_vars` which returns a `NamedTupleLiteral | NilLiteral`. I think `NilLiteral` is more appropriate than `Nop` here because this method doesn't return a direct subnode (i.e. `{{ x.rescues }}` does not reconstruct the original rescue branches, as it's only an array literal).

I have no strong opinion on whether we should return an empty array here. (The underlying array is nilable presumably for performance reasons.) The macro API is rather inconsistent in this aspect; `Call#named_args` uses `| Nop`, `ProcNotation#inputs` returns an empty array, and then there's `Generic#type_vars` as mentioned.